scale-decode/src/visitor/decode.rs

```diff
@@ -12,329 +12,343 @@
-        // Pull an array from the data if we can, preserving the lifetime.
-        let arr: &'scale [u8; 32] = match (*data).try_into() {
-            Ok(arr) => arr,
-            Err(_) => return Err(DecodeError::NotEnoughInput),
-        };
-        // If this works out, remember to shift data 32 bytes forward.
-        *data = &data[32..];
-        Ok(arr)
+        res
     }
-    match ty {
-        TypeDefPrimitive::Bool => {
-            err_if_compact!(is_compact, ty);
-            let b = bool::decode(data).map_err(|e| e.into())?;
-            visitor.visit_bool(b, ty_id)
+
+    fn visit_tuple<TypeIds>(self, type_ids: TypeIds) -> Self::Value
+    where
+        TypeIds: ExactSizeIterator<Item = &'resolver Self::TypeId>,
+    {
+        // guard against invalid compact types: only composites with 1 field can be compact encoded
+        if self.is_compact && type_ids.len() != 1 {
+            return Err(DecodeError::CannotDecodeCompactIntoType.into());
         }
-        TypeDefPrimitive::Char => {
-            err_if_compact!(is_compact, ty);
-            // Treat chars as u32's
-            let val = u32::decode(data).map_err(|e| e.into())?;
-            let c = char::from_u32(val).ok_or(DecodeError::InvalidChar(val))?;
-            visitor.visit_char(c, ty_id)
+
+        let mut fields = type_ids.map(Field::unnamed);
+        let mut items = Tuple::new(self.data, &mut fields, self.types, self.is_compact);
+        let res = self.visitor.visit_tuple(&mut items, self.type_id);
+
+        // Skip over any bytes that the visitor chose not to decode:
+        items.skip_decoding()?;
+        *self.data = items.bytes_from_undecoded();
+
+        res
+    }
+
+    fn visit_primitive(self, primitive: Primitive) -> Self::Value {
+        macro_rules! err_if_compact {
+            ($is_compact:expr) => {
+                if $is_compact {
+                    return Err(DecodeError::CannotDecodeCompactIntoType.into());
+                }
+            };
         }
-        TypeDefPrimitive::Str => {
-            err_if_compact!(is_compact, ty);
-            // Avoid allocating; don't decode into a String. instead, pull the bytes
-            // and let the visitor decide whether to use them or not.
-            let mut s = Str::new(data)?;
-            // Since we aren't decoding here, shift our bytes along to after the str:
-            *data = s.bytes_after();
-            visitor.visit_str(&mut s, ty_id)
+
+        fn decode_32_bytes<'scale>(
+            data: &mut &'scale [u8],
+        ) -> Result<&'scale [u8; 32], DecodeError> {
+            // Pull an array from the data if we can, preserving the lifetime.
+            let arr: &'scale [u8; 32] = match (*data).try_into() {
+                Ok(arr) => arr,
+                Err(_) => return Err(DecodeError::NotEnoughInput),
+            };
+            // If this works out, remember to shift data 32 bytes forward.
```

Comment:
I would tweak this comment a bit because it's a bit confusing.

```suggestion
            // If the bytes were read successfully mark them as read and advance the pointer to read
            // the next bytes in the scale bytes buffer.
```

Reply:
alternatively, a  helper function to move the pointer this for would also be make it clear to me ^^

Reply:
I just tweaked the comment; we're already in a helper function so I didn't want to make another :D